src/kaleidoscope/EventQueue.h

```diff
@@ -0,0 +1,105 @@
+// -*- mode: c++ -*-
+/* Kaleidoscope - Firmware for computer input devices
+ * Copyright (C) 2013-2019  Keyboard.io, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <Arduino.h>
+
+#include "kaleidoscope/Kaleidoscope.h"
+#include "kaleidoscope/KeyAddr.h"
+
+namespace kaleidoscope {
+
+// This class defines a generic event queue that stores both key press and
+// release events, recording the key address, a timestamp, and the keyswitch
+// state (press or release).
+template <uint8_t _max_length,
+          typename _Bitfield  = uint8_t,
+          typename _Timestamp = uint16_t>
+class EventQueue {
+
+  static_assert(_max_length <= 255,
```

Comment:
Restricting `_max_length` by `255` does not make sense as it is restricted by `64` anyway, given that `uint64_t` is the largest possible unsigned integer datatype to be used for `_Bitfield`.

Reply:
Good point. I'll fix that. It would be crazy to have a queue even that long, anyway.

Reply:
On further consideration, it is _possible_ for `_max_length` to be greater than 64 — if `_Bitfield` is not merely an integer type. That's an academic point here, though, unless we can think of an actual use case for a queue that long.

Reply:
Oh — but if `_Bitfield` isn't just an integer, `bitRead()` won't work as expected…